### PR TITLE
Grpc Status Management

### PR DESCRIPTION
**Pull request by ChristopherDavenport**

### CODE REVIEW

File: core/src/main/scala/org/http4s/grpc/ServerGrpc.scala [hunk 1]
```diff
@@ -105,24 +107,25 @@ object ServerGrpc {
...
       } yield {
         val body = Stream.eval(f(codecs.Messages.decode(decode)(req.body), req.headers))
           .flatMap(codecs.Messages.encodeSingle(encode)(_))
           .through(timeoutStream(_)(timeout.map(_.duration)))
           .onFinalizeCase{
-            case Resource.ExitCase.Errored(_: TimeoutException) => status.set(4)
-            case Resource.ExitCase.Errored(_) => status.set(2)
-            case Resource.ExitCase.Canceled => status.set(1)
+            case Resource.ExitCase.Errored(_: TimeoutException) => status.set((4, None))
+            case Resource.ExitCase.Errored(e) => status.set((2, e.getMessage().some))
```
Comment (armanbilge) [inline core/src/main/scala/org/http4s/grpc/ServerGrpc.scala:124]: I wonder if `toString` is better, that way it also includes the class of the exception in addition to the message. I think :)

```suggestion
            case Resource.ExitCase.Errored(e) => status.set((2, e.toString().some))
```

File: core/src/main/scala/org/http4s/grpc/ServerGrpc.scala [hunk 2]
```diff
@@ -146,24 +149,25 @@ object ServerGrpc {
...
 
         val body = f(codecs.Messages.decode(decode)(req.body), req.headers)
           .through(codecs.Messages.encode(encode))
           .through(timeoutStream(_)(timeout.map(_.duration)))
           .onFinalizeCase{
-            case Resource.ExitCase.Errored(_: TimeoutException) => status.set(4)
-            case Resource.ExitCase.Errored(_) => status.set(2)
-            case Resource.ExitCase.Canceled => status.set(1)
+            case Resource.ExitCase.Errored(_: TimeoutException) => status.set((4, None))
+            case Resource.ExitCase.Errored(e) => status.set((2, e.getMessage().some))
```
Comment (armanbilge) [inline core/src/main/scala/org/http4s/grpc/ServerGrpc.scala:166]:
```suggestion
            case Resource.ExitCase.Errored(e) => status.set((2, e.toString().some))
```

File: core/src/main/scala/org/http4s/grpc/ServerGrpc.scala [hunk 3]
```diff
@@ -146,24 +149,25 @@ object ServerGrpc {
...
+            case Resource.ExitCase.Canceled => status.set((1, None))
             case _ => ().pure[F]
           }
+          .handleErrorWith(_ => Stream.empty) // ensures body closure without rst-stream
```
Comment (armanbilge) [inline core/src/main/scala/org/http4s/grpc/ServerGrpc.scala:170]:
```suggestion
          .mask // ensures body closure without rst-stream
```

File: core/src/main/scala/org/http4s/grpc/ServerGrpc.scala [hunk 4]
```diff
@@ -24,24 +24,25 @@ object ServerGrpc {
...
         val body = Stream.eval(codecs.Messages.decodeSingle(decode)(req.body))
           .evalMap(f(_, req.headers))
           .flatMap(codecs.Messages.encodeSingle(encode)(_))
           .through(timeoutStream(_)(timeout.map(_.duration)))
           .onFinalizeCase{
-            case Resource.ExitCase.Errored(_: TimeoutException) => status.set(4)
-            case Resource.ExitCase.Errored(_) => status.set(2)
-            case Resource.ExitCase.Canceled => status.set(1)
+            case Resource.ExitCase.Errored(_: TimeoutException) => status.set((4, None))
+            case Resource.ExitCase.Errored(e) => status.set((2, e.getMessage().some))
```
Comment (armanbilge) [inline core/src/main/scala/org/http4s/grpc/ServerGrpc.scala:41]:
```suggestion
            case Resource.ExitCase.Errored(e) => status.set((2, e.toString().some))
```

**Review by armanbilge — APPROVED** [review]
Minor comments, take-or-leave

File: core/src/main/scala/org/http4s/grpc/ServerGrpc.scala (hunk 1 above)
Comment (ChristopherDavenport) [reply core/src/main/scala/org/http4s/grpc/ServerGrpc.scala:124]: I went with what fs2-grpc was returning. I dont mind a different choice though.

**Review by armanbilge — APPROVED** [review]

**Pull request merged**